x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteClusterPermissionGroup.java

```diff
@@ -107,15 +106,12 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         RemoteClusterPermissionGroup that = (RemoteClusterPermissionGroup) o;
-        return Arrays.equals(clusterPrivileges, that.clusterPrivileges)
-            && Arrays.equals(remoteClusterAliases, that.remoteClusterAliases)
-            && Objects.equals(remoteClusterAliasMatcher, that.remoteClusterAliasMatcher);
```

Comment:
I had to remove `remoteClusterAliasMatcher` from both hashCode and equals implementation as it is always wrong due to the nature of `Predicate`. I'm thinking of not overriding them in `StringMatcher` to avoid confusion.
 
```java
var first = StringMatcher.of("remote-*");
var second = StringMatcher.of("remote-*");
first.equals(second); // this will never be true because of `Predicate` in `StringMatcher`
```

Reply:
weird ... the unit tests extend AbstractXContentSerializingTestCase which in turn tests org.elasticsearch.test.AbstractWireTestCase#testConcurrentEquals.. which should catch things like.

Taking a deeper look and adding this test:
>        assertTrue(StringMatcher.of("*").equals(StringMatcher.of("*"))); //passes
>        assertTrue(StringMatcher.of("foo").equals(StringMatcher.of("foo")));//fails

So those equality tests passed since i guess '*' is a special case.  

I explicitly added the equals/hash to StringMatcher for this purpose (but only relied on tests from AbstractXContentSerializingTestCase of RemoteClusterPermissionGroup). Can you also remove the equals and hash code, or better update them to always return false with a comment to why (that is super trappy behavior). 
   
Also, if you remove this, can you add a comment we explicitly omit the matcher ? It doesn't need be there and the only reason I included it in the equality was to avoid messages like this since most devs will just let the IDE generate these.

Also thanks for fixing the missing RoleDescriptor equality. I just ran the failing RoleDescriptor test a 1000 times and it only failed 3/1000 of the times. We really shouldn't use that randomization :/

Reply:
> Can you also remove the equals and hash code, or better update them to always return false with a comment to why (that is super trappy behavior).

Returning false would be wrong when the same object is compared. So I decided to remove them since this is the correct behaviour.

Reply:
>Also thanks for fixing the missing RoleDescriptor equality. I just ran the failing RoleDescriptor test a 1000 times and it only failed 3/1000 of the times (and got it to pass with 0/1000 failures) . We really shouldn't use that much randomization :/

This was accidental. I only caught it because `isEmpty` was failing, which then uncovered equals issue when reloading `FileRolesStore`.